lib/managers/webhooks.js

```diff
@@ -193,6 +193,120 @@ Webhooks.prototype.delete = function(webhookID, callback) {
+		return undefined;
+	}
+
+	let hmac = crypto.createHmac('sha256', signatureKey);
+	hmac.update(body);
+	hmac.update(headers['box-delivery-timestamp']);
+
+	const signature = hmac.digest('base64');
+	// console.log('Signature: %s', signature);
+
+	return signature;
+}
+
+/**
+ * Validate the message signature (see https://docs.box.com/reference#signatures)
+ *
+ * @param {string} body - The request body of the webhook message
+ * @param {Object} headers - The request headers of the webhook message
+ * @param {string} primarySignatureKey - The primary signature to verify the message with
+ * @param {?string} secondarySignatureKey - The secondary signature to verify the message with
+ * @returns {boolean} - True or false
+ * @private
+ */
+function validateSignature(body, headers, primarySignatureKey, secondarySignatureKey) {
+	// Either the primary or secondary signature must match the corresponding signature from Box
+	// (The use of two signatures allows the signing keys to be rotated one at a time)
+	const primarySignature = computeSignature(body, headers, primarySignatureKey);
+
+	if (primarySignature && primarySignature === headers['box-signature-primary']) {
+		// console.log('Primary signature verified');
+		return true;
+	}
+
+	const secondarySignature = computeSignature(body, headers, secondarySignatureKey);
+
+	if (secondarySignature && secondarySignature === headers['box-signature-secondary']) {
+		// console.log('Secondary signature verified');
+		return true;
+	}
+
+	// console.log('Signature not verified');
+	return false;
+}
+
+/**
+ * Validate that the delivery timestamp is not too far in the past (to prevent replay attacks)
+ *
+ * @param {Object} headers - The request headers of the webhook message
+ * @param {int} maxMessageAge - The maximum message age (in seconds)
+ * @returns {boolean} - True or false
+ * @private
+ */
+function validateDeliveryTimestamp(headers, maxMessageAge) {
+	const deliveryTime = Date.parse(headers['box-delivery-timestamp']);
+	const currentTime = Date.now();
+	const messageAge = (currentTime - deliveryTime) / 1000;
+
+	// console.log('Message age: %d', messageAge);
+
+	if (messageAge > maxMessageAge) {
+		// console.log('Message too old');
+		return false;
+	}
+
+	return true;
+}
+
+// This prevents replay attacks
+const maxMessageAge = 10 * 60; // 10 minutes
+
+/**
+ * Validate a webhook message by verifying the signature and the delivery timestamp
+ *
+ * @param {string} body - The request body of the webhook message
+ * @param {Object} headers - The request headers of the webhook message
+ * @param {string} primarySignatureKey - The primary signature to verify the message with
+ * @param {?string} secondarySignatureKey - The secondary signature to verify the message with
+ * @returns {boolean} - True or false
+ */
+Webhooks.validateMessage = function(body, headers, primarySignatureKey, secondarySignatureKey) {
```

Comment:
We might want to consider making the timestamp delta configurable; we've run into issues before where clients' system clocks are way out of sync with Box